File: backends/quarkus-java/src/main/java/dev/stackverse/backend/AdminService.java
package dev.stackverse.backend;

import com.fasterxml.jackson.databind.ObjectMapper;
import io.quarkus.security.identity.SecurityIdentity;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.ws.rs.core.Response;
import java.time.Instant;
import java.time.LocalDate;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import javax.sql.DataSource;
import org.eclipse.microprofile.jwt.JsonWebToken;
import org.jboss.logging.Logger;

@ApplicationScoped
public class AdminService extends ServiceSupport {
    @Inject
    public AdminService(
            DataSource dataSource,
            JsonWebToken jwt,
            SecurityIdentity securityIdentity,
            ObjectMapper mapper,
            Localizer localizer) {
        super(dataSource, jwt, securityIdentity, mapper, localizer);
    }

    public Response listUsers(RequestContext request) {
        requireRole("admin");
        int page = pagingPage(request);
        int size = pageSize(request);
        String q = singleParam(request, "q");
        maxLength(q, 100, "q");
        String status = singleParam(request, "status");
        if (status != null && !Set.of("active", "blocked").contains(status)) {
            throw StackverseProblem.badRequest("status must be one of: active, blocked");
        }
        PageResponse<UserAccountResponse> body =
                withConnection(
                        connection -> {
                            SqlWhere where = new SqlWhere();
                            if (q != null && !q.isBlank()) {
                                where.and(
                                        "u.username ilike ? escape '\\'",
                                        "%" + escapeLike(q) + "%");
                            }
                            if (status != null) {
                                where.and("u.status = ?", status);
                            }
                            long total =
                                    scalarLong(
                                            connection,
                                            "select count(*) from user_accounts u " + where.sql(),
                                            where.params());
                            List<Object> params = new ArrayList<>(where.params());
                            params.add(size);
                            params.add(offset(page, size));
                            List<UserAccountResponse> items =
                                    query(
                                            connection,
                                            userAccountSelect()
                                                    + " "
                                                    + where.sql()
                                                    + " order by u.last_seen desc, u.username asc limit ? offset ?",
                                            params,
                                            rs -> userAccountResponse(userAccount(rs)));
                            return pageResponse(items, page, size, total);
                        });
        return Response.ok(body).build();
    }

    public Response getUser(String username) {
        requireRole("admin");
        UserAccount account =
                withConnection(connection -> findUserAccount(connection, username))
                        .orElseThrow(StackverseProblem::notFound);
        return Response.ok(userAccountResponse(account)).build();
    }

    public Response setUserStatus(String username, UserStatusInput input) {
        Caller caller = requireRole("admin");
        if (!Set.of("active", "blocked").contains(input.status())) {
            throw StackverseProblem.badRequest("status is required");
        }
        if ("blocked".equals(input.status())) {
            if (username.equals(caller.username())) {
                throw StackverseProblem.conflict("Admins cannot block themselves.");
            }
        }
        inTransaction(
                connection -> {
                    queryOne(
                                    connection,
                                    "select username from user_accounts where username = ? for update",
                                    List.of(username),
                                    rs -> rs.getString("username"))
                            .orElseThrow(StackverseProblem::notFound);
                    if ("blocked".equals(input.status())) {
                        execute(
                                connection,
                                "update user_accounts set status = 'blocked', blocked_reason = ? where username = ?",
                                params(input.reason(), username));
                        recordAudit(
                                connection,
                                caller.username(),
                                "user.blocked",
                                "user",
                                username,
                                Map.of("reason", input.reason()));
                    } else {
                        execute(
                                connection,
                                "update user_accounts set status = 'active', blocked_reason = null where username = ?",
                                List.of(username));
                        recordAudit(
                                connection,
                                caller.username(),
                                "user.unblocked",
                                "user",
                                username,
                                null);
                    }
                    return null;
                });
        StackverseLog.event(
                LOG,
                Logger.Level.INFO,
                "blocked".equals(input.status()) ? "user_blocked" : "user_unblocked",
                "success",
                "blocked".equals(input.status())
                        ? "User account blocked"
                        : "User account unblocked",
                Map.of(
                        "actor",
                        caller.username(),
                        "resource_type",
                        "user",
                        "resource_id",
                        username));
        UserAccount account =
                withConnection(connection -> findUserAccount(connection, username))
                        .orElseThrow(StackverseProblem::notFound);
        return Response.ok(userAccountResponse(account)).build();
    }

    public Response auditLog(RequestContext request) {
        requireRole("admin");
        int page = pagingPage(request);
        int size = pageSize(request);
        PageResponse<AuditResponse> body =
                withConnection(
                        connection -> {
                            SqlWhere where = new SqlWhere();
                            equalFilter(request, where, "actor", "actor");
                            equalFilter(request, where, "action", "action");
                            equalFilter(request, where, "target_type", "targetType");
                            equalFilter(request, where, "target_id", "targetId");
                            Instant from = timeParam(request, "from");
                            Instant to = timeParam(request, "to");
                            if (from != null) {
                                where.and("created_at >= ?", from);
                            }
                            if (to != null) {
                                where.and("created_at <= ?", to);
                            }
                            long total =
                                    scalarLong(
                                            connection,
                                            "select count(*) from audit_entries " + where.sql(),
                                            where.params());
                            List<Object> params = new ArrayList<>(where.params());
                            params.add(size);
                            params.add(offset(page, size));
                            List<AuditResponse> items =
                                    query(
                                            connection,
                                            "select * from audit_entries "
                                                    + where.sql()
                                                    + " order by created_at desc, id desc limit ? offset ?",
                                            params,
                                            rs -> auditResponse(audit(rs)));
                            return pageResponse(items, page, size, total);
                        });
        return Response.ok(body).build();
    }

    public Response stats(RequestContext request) {
        requireRole("moderator");
        Map<String, Object> body =
                withConnection(
                        connection -> {
                            Map<String, Object> totals = new LinkedHashMap<>();
                            totals.put(
                                    "users",
                                    scalarLong(
                                            connection,
                                            "select count(*) from user_accounts",
                                            List.of()));
                            totals.put(
                                    "bookmarks",
                                    scalarLong(
                                            connection,
                                            "select count(*) from bookmarks",
                                            List.of()));
                            totals.put(
                                    "publicBookmarks",
                                    scalarLong(
                                            connection,
                                            "select count(*) from bookmarks where visibility = 'public'",
                                            List.of()));
                            totals.put(
                                    "hiddenBookmarks",
                                    scalarLong(
                                            connection,
                                            "select count(*) from bookmarks where status = 'hidden'",
                                            List.of()));
                            totals.put(
                                    "openReports",
                                    scalarLong(
                                            connection,
                                            "select count(*) from reports where status = 'open'",
                                            List.of()));

                            LocalDate today = LocalDate.now(ZoneOffset.UTC);
                            LocalDate from = today.minusDays(29);
                            Map<LocalDate, Long> bookmarksCreated =
                                    countPerDay(connection, "bookmarks", "created_at", from);
                            Map<LocalDate, Long> activeUsers =
                                    countPerDay(connection, "user_accounts", "last_seen", from);
                            List<Map<String, Object>> daily = new ArrayList<>();
                            for (int i = 0; i < 30; i++) {
                                LocalDate date = from.plusDays(i);
                                Map<String, Object> row = new LinkedHashMap<>();
                                row.put("date", date.toString());
                                row.put(
                                        "bookmarksCreated",
                                        bookmarksCreated.getOrDefault(date, 0L));
                                row.put("activeUsers", activeUsers.getOrDefault(date, 0L));
                                daily.add(row);
                            }
                            List<Map<String, Object>> topTags =
                                    query(
                                            connection,
                                            "select tag, count(*) as count from bookmarks, unnest(tags) as tag"
                                                    + " group by tag order by count desc, tag asc limit 10",
                                            List.of(),
                                            rs -> {
                                                Map<String, Object> row = new LinkedHashMap<>();
                                                row.put("tag", rs.getString("tag"));
                                                row.put("count", rs.getLong("count"));
                                                return row;
                                            });
                            Map<String, Object> response = new LinkedHashMap<>();
                            response.put("totals", totals);
                            response.put("daily", daily);
                            response.put("topTags", topTags);
                            return response;
                        });
        return etag(request, body, null);
    }
}

File: backends/scala-http4s/src/main/scala/dev/stackverse/http4s/AdminService.scala
package dev.stackverse.http4s

import cats.effect.IO
import io.circe.{Json, JsonObject}
import org.http4s.*

import java.sql.{Connection, SQLException}
import java.time.{Instant, LocalDate, ZoneOffset}
import java.util.UUID
import scala.collection.mutable.ArrayBuffer
import scala.util.Try

trait AdminOperations {
  def listUsers(req: Request[IO]): IO[Response[IO]]
  def getUser(req: Request[IO], username: String): IO[Response[IO]]
  def setUserStatus(req: Request[IO], username: String): IO[Response[IO]]
  def auditLog(req: Request[IO]): IO[Response[IO]]
  def stats(req: Request[IO]): IO[Response[IO]]
}

final class AdminService(db: Db, auth: AuthService, logger: EventLogger, repository: SharedRepository)
    extends AdminOperations {
  import RouteSupport.*
  import Wire.*
  import repository.*

  override def listUsers(req: Request[IO]): IO[Response[IO]] = IO.blocking {
    auth.requireRole(req, "admin")
    val (page, size) = paging(query(req))
    val q = single(query(req), "q")
    maxLength(q, 100, "q")
    val status = single(query(req), "status")
    if (status.exists(value => value != "active" && value != "blocked"))
      throw BadRequestProblem(s"unknown status: ${status.get}")
    val clauses = ArrayBuffer("true")
    val params = ArrayBuffer.empty[Any]
    q.filter(_.trim.nonEmpty).foreach { value =>
      clauses += "u.username ilike ? escape '\\'"
      params += s"%${escapeLike(value)}%"
    }
    status.foreach { value =>
      clauses += "u.status = ?"
      params += value
    }
    val where = clauses.mkString(" and ")
    val payload = db.withConnection { conn =>
      val rows = db.query(
        conn,
        s"$withBookmarkCountQuery where $where order by u.last_seen desc, u.username asc limit ? offset ?",
        params.toSeq ++ Seq(size, page * size)
      )(Rows.user)
      val total = count(conn, s"select count(*) as count from user_accounts u where $where", params.toSeq)
      pagePayload(rows.map(Responses.user), page, size, total)
    }
    jsonResponse(Status.Ok, payload)
  }

  override def getUser(req: Request[IO], username: String): IO[Response[IO]] = IO.blocking {
    auth.requireRole(req, "admin")
    val row = db.withConnection(conn => findUser(conn, username).getOrElse(throw NotFoundProblem()))
    jsonResponse(Status.Ok, Responses.user(row))
  }

  override def setUserStatus(req: Request[IO], username: String): IO[Response[IO]] =
    for {
      body <- jsonBody(req)
      response <- IO.blocking {
        val caller = auth.requireRole(req, "admin")
        val status = body("status").flatMap(_.asString).getOrElse(throw BadRequestProblem("status is required"))
        if (status != "active" && status != "blocked") throw BadRequestProblem("status is required")
        val reason = body("reason").flatMap(_.asString).map(_.trim)
        if (status == "blocked") {
          val validator = Validator()
          validator.check(reason.exists(_.nonEmpty), "reason", "validation.block.reason.required")
          validator.check(reason.forall(_.length <= 1000), "reason", "validation.block.reason.too-long")
          validator.throwIfInvalid()
          if (username == caller.username) throw ConflictProblem("Admins cannot block themselves.")
        }
        db.transaction { conn =>
          db.one(conn, "select username from user_accounts where username = ? for update", Seq(username))(_ => true)
            .getOrElse(throw NotFoundProblem())
          if (status == "blocked") {
            db.execute(
              conn,
              "update user_accounts set status = 'blocked', blocked_reason = ? where username = ?",
              Seq(reason, username)
            )
            recordAudit(
              conn,
              caller.username,
              "user.blocked",
              "user",
              username,
              JsonObject("reason" -> reason.fold(Json.Null)(Json.fromString))
            )
          } else {
            db.execute(
              conn,
              "update user_accounts set status = 'active', blocked_reason = null where username = ?",
              Seq(username)
            )
            recordAudit(conn, caller.username, "user.unblocked", "user", username)
          }
        }
        logger.event(
          "info",
          if (status == "blocked") "user_blocked" else "user_unblocked",
          "success",
          if (status == "blocked") "User account blocked" else "User account unblocked",
          "actor" -> Json.fromString(caller.username),
          "resource_type" -> Json.fromString("user"),
          "resource_id" -> Json.fromString(username)
        )
        val row = db.withConnection(conn => findUser(conn, username).getOrElse(throw NotFoundProblem()))
        jsonResponse(Status.Ok, Responses.user(row))
      }
    } yield response

  override def auditLog(req: Request[IO]): IO[Response[IO]] = IO.blocking {
    auth.requireRole(req, "admin")
    val (page, size) = paging(query(req))
    val clauses = ArrayBuffer("true")
    val params = ArrayBuffer.empty[Any]
    Seq("actor" -> "actor", "action" -> "action", "targetType" -> "target_type", "targetId" -> "target_id").foreach {
      case (param, column) =>
        single(query(req), param).foreach { value =>
          clauses += s"$column = ?"
          params += value
        }
    }
    single(query(req), "from").foreach { value =>
      clauses += "created_at >= ?"
      params += parseInstantParam(value, "from")
    }
    single(query(req), "to").foreach { value =>
      clauses += "created_at <= ?"
      params += parseInstantParam(value, "to")
    }
    val where = clauses.mkString(" and ")
    val payload = db.withConnection { conn =>
      val rows = db.query(
        conn,
        s"select * from audit_entries where $where order by created_at desc, id desc limit ? offset ?",
        params.toSeq ++ Seq(size, page * size)
      )(Rows.audit)
      val total = count(conn, s"select count(*) as count from audit_entries where $where", params.toSeq)
      pagePayload(rows.map(Responses.audit), page, size, total)
    }
    jsonResponse(Status.Ok, payload)
  }

  override def stats(req: Request[IO]): IO[Response[IO]] = IO.blocking {
    auth.requireRole(req, "moderator")
    val today = LocalDate.now(ZoneOffset.UTC)
    val from = today.minusDays(29).atStartOfDay().toInstant(ZoneOffset.UTC)
    val payload = db.withConnection { conn =>
      val users = count(conn, "select count(*) as count from user_accounts")
      val bookmarks = count(conn, "select count(*) as count from bookmarks")
      val publicBookmarks = count(conn, "select count(*) as count from bookmarks where visibility = 'public'")
      val hiddenBookmarks = count(conn, "select count(*) as count from bookmarks where status = 'hidden'")
      val openReports = count(conn, "select count(*) as count from reports where status = 'open'")
      val createdPerDay = countPerDay(conn, "bookmarks", "created_at", from)
      val activePerDay = countPerDay(conn, "user_accounts", "last_seen", from)
      val topTags = db.query(
        conn,
        """select tag, count(*) as count from bookmarks, unnest(tags) as tag
          |group by tag order by count desc, tag asc limit 10""".stripMargin
      )(rs => Json.obj("tag" -> Json.fromString(rs.getString("tag")), "count" -> Json.fromLong(rs.getLong("count"))))
      val daily = (0 until 30).map { offset =>
        val date = today.minusDays(29 - offset.toLong)
        val bookmarksCreated = createdPerDay.get(date.toString).getOrElse(0L)
        val activeUsers = activePerDay.get(date.toString).getOrElse(0L)
        Json.obj(
          "date" -> Json.fromString(date.toString),
          "bookmarksCreated" -> Json.fromLong(bookmarksCreated),
          "activeUsers" -> Json.fromLong(activeUsers)
        )
      }
      Json.obj(
        "totals" -> Json.obj(
          "users" -> Json.fromLong(users),
          "bookmarks" -> Json.fromLong(bookmarks),
          "publicBookmarks" -> Json.fromLong(publicBookmarks),
          "hiddenBookmarks" -> Json.fromLong(hiddenBookmarks),
          "openReports" -> Json.fromLong(openReports)
        ),
        "daily" -> Json.arr(daily*),
        "topTags" -> Json.arr(topTags*)
      )
    }
    withEtag(req, payload)
  }

}
